File: client/services/videoGeneration.ts
const API_BASE = "http://localhost:8000";
const FAL_API_KEY = process.env.FAL_KEY;
const FAL_API_ENDPOINT = "https://api.fal.ai/v1/queues/fal-ai/veo3.1/reference-to-video/submit";

export interface VideoGenerationRequest {
  prompt: string;
  image_urls: string[];
  duration?: string;
  resolution?: string;
  aspect_ratio?: string;
  generate_audio?: boolean;
}

export interface VideoGenerationResponse {
  status: string;
  video_url?: string;
  message: string;
}

export interface HalftimeVideoRequest {
  reference_image_urls: string[];
  home_team: string;
  away_team: string;
  quarter?: number;
  home_score?: number;
  away_score?: number;
}

/**
 * Generate a video from reference images using Veo 3.1
 * @param prompt - Description of the desired video
 * @param imageUrls - List of reference image URLs
 * @param options - Optional parameters (duration, resolution, etc.)
 * @returns Generated video URL or null on error
 */
export async function generateVideo(
  prompt: string,
  imageUrls: string[],
  options?: {
    duration?: string;
    resolution?: string;
    aspect_ratio?: string;
    generate_audio?: boolean;
  }
): Promise<string | null> {
  try {
    if (!imageUrls || imageUrls.length === 0) {
      console.error("At least one reference image URL is required");
      return null;
    }

    if (!prompt || prompt.trim().length === 0) {
      console.error("Prompt cannot be empty");
      return null;
    }

    const request: VideoGenerationRequest = {
      prompt,
      image_urls: imageUrls,
      duration: options?.duration || "8s",
      resolution: options?.resolution || "720p",
      aspect_ratio: options?.aspect_ratio || "16:9",
      generate_audio: options?.generate_audio !== false,
    };

    console.log("Requesting video generation...", {
      promptLength: prompt.length,
      imageCount: imageUrls.length,
      ...options,
    });

    const response = await fetch(`${API_BASE}/generate_video`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(request),
    });

    if (!response.ok) {
      const error = await response.json();
      console.error("Video generation failed:", error);
      return null;
    }

    const data: VideoGenerationResponse = await response.json();

    if (data.status === "success" && data.video_url) {
      console.log("Video generated successfully:", data.video_url);
      return data.video_url;
    } else {
      console.error("Video generation returned no URL:", data.message);
      return null;
    }
  } catch (error) {
    console.error("Video generation error:", error);
    return null;
  }
}

/**
 * Generate a Super Bowl halftime highlight video using fal-ai Veo 3.1
 * @param referenceImageUrls - List of highlight image URLs (4+ recommended)
 * @param gameContext - Current game state (teams, scores, quarter)
 * @returns Generated video URL or null on error
 */
export async function generateHalftimeVideo(
  referenceImageUrls: string[],
  gameContext: {
    homeTeam: string;
    awayTeam: string;
    quarter?: number;
    homeScore?: number;
    awayScore?: number;
  }
): Promise<string | null> {
  try {
    if (!referenceImageUrls || referenceImageUrls.length === 0) {
      console.error("At least one reference image URL is required");
      return null;
    }

    if (!FAL_API_KEY) {
      console.error("FAL_KEY environment variable not set");
      return null;
    }

    // Create detailed prompt for the video generation
    const prompt = `Create an exciting Super Bowl LIX halftime highlight reel video.

Game Context: ${gameContext.homeTeam} vs ${gameContext.awayTeam}, Quarter ${gameContext.quarter || 2}, Score: ${gameContext.homeTeam} ${gameContext.homeScore || 0} - ${gameContext.awayTeam} ${gameContext.awayScore || 0}

Action: Smoothly transition between key highlight moments from the game, showing epic plays, big catches, and intense moments
Style: Professional NFL broadcast highlights with dynamic camera movements and smooth transitions
Camera motion: Zoom in on key players, pan across field action, follow the ball movement
Ambiance: High-energy stadium atmosphere, dramatic lighting, professional sports broadcast feel
Duration: 8 seconds
Output: 720p video quality with natural motion and realistic animations`;

    console.log("Requesting halftime video generation from fal-ai...", {
      imageCount: referenceImageUrls.length,
      homeTeam: gameContext.homeTeam,
      awayTeam: gameContext.awayTeam,
    });

    const falRequest = {
      prompt,
      image_urls: referenceImageUrls,
      duration: "8s",
      resolution: "720p",
      aspect_ratio: "16:9",
      generate_audio: true,
    };

    const response = await fetch(FAL_API_ENDPOINT, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "Authorization": `Key ${FAL_API_KEY}`,
      },
      body: JSON.stringify(falRequest),
    });

    if (!response.ok) {
      const error = await response.text();
      console.error("fal-ai halftime video generation failed:", error);
      return null;
    }

    const data = await response.json();

    // fal-ai returns a request_id, we need to poll for the result
    if (data.request_id) {
      console.log("Video generation submitted with request_id:", data.request_id);

      // Poll for result (max 30 attempts, 2 second intervals)
      for (let i = 0; i < 30; i++) {
        await new Promise(resolve => setTimeout(resolve, 2000));

        const statusResponse = await fetch(
          `https://api.fal.ai/v1/queues/default/requests/${data.request_id}/status`,
          {
            headers: {
              "Authorization": `Key ${FAL_API_KEY}`,
            },
          }
        );

        if (statusResponse.ok) {
          const statusData = await statusResponse.json();

          if (statusData.status === "COMPLETED" && statusData.output?.video?.url) {
            console.log("Halftime video generated successfully:", statusData.output.video.url);
            return statusData.output.video.url;
          } else if (statusData.status === "FAILED") {
            console.error("Video generation failed:", statusData.error);
            return null;
          }
        }
      }

      console.error("Video generation timeout after polling");
      return null;
    } else {
      console.error("No request_id returned from fal-ai");
      return null;
    }
  } catch (error) {
    console.error("Halftime video generation error:", error);
    return null;
  }
}

/**
 * Check if video generation is available
 * @returns Status object with capabilities
 */
export async function getVideoGenerationStatus() {
  try {
    const response = await fetch(`${API_BASE}/video_generation_status`);

    if (!response.ok) {
      console.error("Failed to get video generation status");
      return null;
    }

    return await response.json();
  } catch (error) {
    console.error("Error checking video generation status:", error);
    return null;
  }
}

File: backend/api/routes/video_generation.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import Optional

from services.veo_service import veo_service
from utils.logger import logger

router = APIRouter()


class VideoGenerationRequest(BaseModel):
    """Request model for video generation."""
    prompt: str
    image_urls: list[str]
    duration: Optional[str] = "8s"
    resolution: Optional[str] = "720p"
    aspect_ratio: Optional[str] = "16:9"
    generate_audio: Optional[bool] = True


class HalftimeVideoRequest(BaseModel):
    """Request model for halftime highlight video generation."""
    reference_image_urls: list[str]
    home_team: str
    away_team: str
    quarter: Optional[int] = 2
    home_score: Optional[int] = 0
    away_score: Optional[int] = 0


class VideoGenerationResponse(BaseModel):
    """Response model for video generation."""
    status: str
    video_url: Optional[str] = None
    message: str


@router.post("/generate_video", response_model=VideoGenerationResponse)
async def generate_video(request: VideoGenerationRequest):
    """
    Generate a video from reference images using Veo 3.1.

    Requires:
    - prompt: Text description of the desired video
    - image_urls: List of reference image URLs (at least 1)

    Optional:
    - duration: Video duration ("8s")
    - resolution: "720p", "1080p", or "4k"
    - aspect_ratio: "16:9" or "9:16"
    - generate_audio: Whether to generate audio

    Returns:
    - status: "success" or "error"
    - video_url: URL to the generated video (if successful)
    """
    try:
        # Validate request
        if not request.image_urls or len(request.image_urls) == 0:
            raise HTTPException(
                status_code=400,
                detail="At least one reference image URL is required"
            )

        if not request.prompt or len(request.prompt.strip()) == 0:
            raise HTTPException(
                status_code=400,
                detail="Prompt cannot be empty"
            )

        # Initialize service if needed
        if not veo_service._initialized:
            if not veo_service.initialize():
                raise HTTPException(
                    status_code=503,
                    detail="Veo service is not available. Please check VEO_API_KEY configuration."
                )

        logger.info(f"Generating video with {len(request.image_urls)} reference images")

        # Generate video
        result = await veo_service.generate_video_from_images(
            prompt=request.prompt,
            image_urls=request.image_urls,
            duration=request.duration,
            resolution=request.resolution,
            aspect_ratio=request.aspect_ratio,
            generate_audio=request.generate_audio,
        )

        if result and result.get("video_url"):
            return VideoGenerationResponse(
                status="success",
                video_url=result["video_url"],
                message=f"Video generated successfully with {result['image_count']} reference images"
            )
        else:
            logger.error("Video generation returned no URL")
            raise HTTPException(
                status_code=500,
                detail="Video generation failed or returned no URL"
            )

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Video generation error: {e}")
        raise HTTPException(
            status_code=500,
            detail=f"Video generation failed: {str(e)}"
        )


@router.post("/generate_halftime_video", response_model=VideoGenerationResponse)
async def generate_halftime_video(request: HalftimeVideoRequest):
    """
    Generate a Super Bowl halftime highlight video.

    Takes 4+ reference images and creates a dynamic halftime video
    featuring the two teams with current game context.

    Requires:
    - reference_image_urls: List of highlight image URLs (4+ recommended)
    - home_team: Home team name
    - away_team: Away team name

    Optional:
    - quarter: Current quarter
    - home_score: Home team score
    - away_score: Away team score

    Returns:
    - status: "success" or "error"
    - video_url: URL to the generated halftime video
    """
    try:
        # Validate minimum images for halftime video
        if len(request.reference_image_urls) < 1:
            raise HTTPException(
                status_code=400,
                detail="At least 4 reference images recommended for halftime video generation"
            )

        # Initialize service if needed
        if not veo_service._initialized:
            if not veo_service.initialize():
                raise HTTPException(
                    status_code=503,
                    detail="Veo service is not available. Please check VEO_API_KEY configuration."
                )

        logger.info(
            f"Generating halftime video: {request.home_team} vs {request.away_team} "
            f"({request.home_score}-{request.away_score}), Q{request.quarter}, "
            f"{len(request.reference_image_urls)} images"
        )

        # Generate halftime video
        result = await veo_service.generate_halftime_video(
            reference_image_urls=request.reference_image_urls,
            game_context={
                "home_team": request.home_team,
                "away_team": request.away_team,
                "quarter": request.quarter,
                "home_score": request.home_score,
                "away_score": request.away_score,
            }
        )

        if result and result.get("video_url"):
            return VideoGenerationResponse(
                status="success",
                video_url=result["video_url"],
                message=f"Halftime video generated successfully"
            )
        else:
            raise HTTPException(
                status_code=500,
                detail="Halftime video generation failed"
            )

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Halftime video generation error: {e}")
        raise HTTPException(
            status_code=500,
            detail=f"Halftime video generation failed: {str(e)}"
        )


@router.get("/video_generation_status")
async def get_video_generation_status():
    """Get information about video generation capabilities."""
    veo_initialized = veo_service.initialize()

    return {
        "veo_enabled": veo_initialized,
        "supported_resolutions": ["720p", "1080p", "4k"],
        "supported_aspect_ratios": ["16:9", "9:16"],
        "video_duration": "8s",
        "recommended_reference_images": 4,
        "features": [
            "Image-to-video generation",
            "Halftime highlight videos",
            "Audio generation",
            "Multiple resolution options",
        ],
    }
